sillsdev/web-languageforge
# Transfer project ownership

Fixes #1560 
Fixes #1561 

## Description

Project owners and site admins will now have an option to transfer ownership of a project to another member of the project. We added tests for this to the unit tests.

We also added a red X icon to the "remove" user option, and added confirm dialogs to both the "make owner" and "remove user" actions.

### Type of Change

Only keep lines below that describe this change, then delete the rest.

- New feature (non-breaking change which adds functionality)

## Screenshots

View of an admin:
![image](https://user-images.githubusercontent.com/56163492/202411763-bcccb26c-0633-49ad-8649-eb4125f06db2.png)

After making someone else owner:
![image](https://user-images.githubusercontent.com/56163492/202412060-6e5ea9e5-11cd-4df8-8f0f-46388a1ccc81.png)


View of an owner (non-admin):
![image](https://user-images.githubusercontent.com/56163492/202412257-dad2d729-20b6-4678-b458-1185d463f5e5.png)

Confirm dialog ("chicken page" 🐔 ):
![image](https://user-images.githubusercontent.com/56163492/202608336-fe97b82b-f208-4ed7-ada2-ffdc67e620c9.png)

Transferred ownership back to "admin"; someone who is not an owner or admin will not see the option to transfer ownership:
![image](https://user-images.githubusercontent.com/56163492/202411361-3bc751a3-9b8f-4f8f-a9cd-9c6e0d105e0f.png)

Cannot transfer to an invitee:
![image](https://user-images.githubusercontent.com/56163492/203212795-37053b25-d8f8-4f6d-9c88-07237dd402e1.png)

Confirm dialog for removing a user:
![image](https://user-images.githubusercontent.com/56163492/202608421-43b86ab1-38a7-4901-b999-4b892367500a.png)

Here you can see that this admin is not the owner of the project:
![image](https://user-images.githubusercontent.com/56163492/202659638-d4714926-14a4-4276-a770-ba987ccef661.png)

but the admin is still shown the delete tab:
![image](https://user-images.githubusercontent.com/56163492/202659694-3d9dc905-1711-4ed7-b80e-498b11d7901c.png)

which the admin uses and then the project is deleted:
![image](https://user-images.githubusercontent.com/56163492/202659802-f57fea04-f5e8-4dd7-aa9b-b28c03b12048.png)



## Checklist


## How to test

I. As an owner, share to another user. Verify that until the other user signs in, you cannnot make them a project owner. Make the other user a manager. Sign in as the other user. This user (a manager) should not be given the option to transfer project ownership. This user should not be shown the delete tab in project settings.
2. Now, as the owner again, make the other user an owner. Sign in as the other user. Verify that this new owner can transfer ownership as well. 
3. Now sign in as an admin and make/view a project. Verify that you can see the project delete tab. Then share to another user. Verify that you can't transfer project ownership before the other user signs in an joins the project. After the user signs in and joins the project, verify that you (an admin) can transfer project ownership to them.

## qa.languageforge.org testing

Testers should add his/her findings to end of the PR in a comment and include screenshots, files, etc that are beneficial.
